### ISSUE TITLE
Extend functionality for unique/duplicated

### ISSUE THREAD

**Issue by braaannigan** [pola-rs/polars#5590]
### Problem description

It would help to provide some more functionality around finding and keeping/dropping unique and duplicated rows:

**Finding**
- `df.is_duplicated()` exists but it would be good to be able to provide a subset
- `df.is_unique()` exists but it would be good to be able to provide a subset

**Keeping/dropping**
- `df.unique(subset=...)` Exists. Drops duplicate rows from this DataFrame. Could perhaps be aliased by drop_duplicates
- `df.duplicated(subset=...)`  Doesn't exist. Drops unique rows from this DataFrame 

The names of `unique` and `duplicated` could do with a verb to show what action they do e.g. drop_duplicated or drop_unique.

**Comment by ritchie46** [comment]
I agree on the second ones as they are harder to express without the argument.

The first ones seems like just another way of doing it and having extra code-complexity. We can write:

```python
   df.select(subset).is_duplicated()
```

**Comment by JulianCologne** [comment]
hi, I brought up this issue in discord so I thought I quickly state my motivation here.
Given the following data I would like to get all rows with duplicated `article` & `amount` and compare the prices for different countries.

```python
import polars as pl

df = pl.DataFrame({
    "article": ["A", "A", "A", "A", "B", "C", "C", "C", "C"],
    "amount": [1, 5, 1, 5, 1, 1, 1, 1, 5],
    "country": ["X", "X", "Y", "Y", "Y", "X", "Y", "Z", "Z"],
    "price": [5.0, 4.0, 5.5, 4.0, 9.0, 1.5, 1.6, 1.7, 1.3]
})

# Your solition works, but is not very efficient/elegant.
df.filter(
    df.select(["article", "amount"]).is_duplicated()
)

# I initially came up with a similar solution but than realized that it is not working as expected.
df.filter(
    pl.col(['article', 'amount']).is_duplicated()  # ??
    # is this evaluated as `pl.col('article').is_duplicated() & pl.col('amount').is_duplicated()` ??
)

# It would be nice to just be able to do this
df.duplicated(subset=['article', 'amount'], keep='all')
```


#  Differences between Pandas and Polars

pandas `duplicated` by default keeps the first occurence of a duplicate and has a `keep` parameter which is very usefull and can be set to `first`, `last` or `False`.
Having a different default behaviour is a little unintuitive and can lead to bugs. Also would be nice to have a `keep` parameter in Polars.

```python
pd.DataFrame({'vals': [1, 2, 3, 1]}).duplicated()
0    False <<<<<
1    False
2    False
3     True

pl.DataFrame({'vals': [1, 2, 3, 1]}).is_duplicated()
[
	true <<<<<
	false
	false
	true
]
```

# `unique` Problems

## inconsistent behaviour of `unique` and `is_unique`
if `is_unique` tells me the values are unique, I would expect `unique` to return only the values that are `true` which is does not do. Indeed it just de-duplicates which is different is my eyes! The value `1` is not unique and should not be returned 

```python
pl.DataFrame({'vals': [1, 2, 3, 1]}).is_unique()
[
	false
	true
	true
	false
]

pl.DataFrame({'vals': [1, 2, 3, 1]}).unique()
┌──────┐
│ vals │
│ ---  │
│ i64  │
╞══════╡
│ 1    │
├╌╌╌╌╌╌┤
│ 2    │
├╌╌╌╌╌╌┤
│ 3    │
└──────┘
```

## `drop_duplicates` instead of `unique`

I my opinion the naming of the `unique` function is not very intuitive and maybe even incorrect.
I would expect `unique` to return only values which are actually UNIQUE.
In practice what is does is de-duplicate the values.
A good solution in my opinion would be to rename `unique` to `drop_duplicates` and add an additional `keep` parameter and add a `unique` method that actually return only elements that are unique (just like `is_unique`).

**Comment by JulianCologne** [comment]
@mcrumiller I can see your point but the argument "because everybody else does it" is not always the best decision I think.

## `unique`
The challenge I see is that `unique` can have two meanings:
1. return elements that are unique in the source
2. return elements that are unique in the destination (deduplication)

Problem is that `is_unique` uses 1) and `unique` uses 2).
This leads to the unexpected behavior I mentioned above:
- given `[1, 2, 3, 1]`
- `is_unique` tells me that 2 and 3 are unique (True) but
- `unique` returns `[1, 2, 3]`

### Possible solution:
`unique` already has `keep=first|last` to specify which element to keep in case of duplicates. We could add `keep='none'` to keep no duplicates.

- `keep='first'` -> `[1, 2, 3]`
- `keep='last'` -> `[2, 3, 1]`
- `keep='none'` -> `[2, 3]`

I think this would be a great addition wihtout breaking/changing any default behaviour! =)

I would also suggest adding `keep` and `subset` to `is_unique` (and `is_duplicated`) to make it more flexible and consistent with `unique`.


## General unique/duplicate use cases
In general looking at unique/duplicate it would be great if polars would cover the following use cases (not talking about methods names but the functionality):

1) **unique**: elements that are unique in the source
2) **distinct**: deduplication/drop-duplicates/distinct elements in the source (pyspark has a `distinct` method; SQL has `DISTINCT` keyword)
3) **duplicate**: duplicate elements in the source

`unique` already covers 2) and with my proposed `keep='none'` would also cover 1).

This still leaves 3) which is quite a common use case (see my example above). I think this would be a great addition to polars!
Suggtested syntax:
```python
df.duplicated(subset=subset, keep='first|last|all')
```

Example
- [1, 2, 3, 4, 3, 2, 1]
- `keep='first'` -> `[1, 2, 3]`
- `keep='last'` -> `[3, 2, 1]`
- `keep='all'` -> `[1, 2, 3, 3, 2, 1]`

---

Any thoughts on
- different behavior of pandas `duplicated` and polars `is_duplicated`? Maybe add the `keep` parameter (with `keep=first/last/none`) to polars?
- adding `subset` parameter to `is_unique` and `is_duplicated`? Very useful especially when filtering with `col(subset).isXX` does not work.
- inside `df.filter()`, `df.select(subset).is_duplicated` works as expected, but `df.col(subset).is_duplicated` is different (see my example above). Is this intended?

**Comment by mcrumiller** [comment]
> @mcrumiller I can see your point but the argument "because everybody else does it" is not always the best decision I think.

I agree that it's not always the best decision, but in this case I think it is. `unique` is an incredibly standard function, and changing it would be akin to reversing the gas and brakes on your car; the new way might be "better" but it's going to confuse almost everyone in the industry.

Also--there are many use cases for finding all of the unique elements in a set, most stemming from avoiding performing repeated operations on the same element. I can't think of any use case for finding all non-duplicated elements, except for very specific problems.

I like the idea of adding more `keep` parameter options, but I absolutely do *not* think that the definition of the `unique` function should deviate from the standard practice.

**Comment by mcrumiller** [comment]
@ritchie46 how would one filter such that the result corresponds to the usual usage of `unique`, i.e. return True for the first time each value is encountered, and False otherwise? In pandas it's done with `~df.duplicated(keep="first")`. Do we have something comparable? I can't find anything in the API but that would be super useful (to identify first instance of each value).
